TopMovies/Core/State/PaginationState.swift

```diff
@@ -0,0 +1,117 @@
+//
+//  PaginatedState.swift
+//  TopMovies
+//
+//  Created by Macbook Pro  on 19.02.2021.
+//
+
+import ReSwift
+
+enum EmptyRequestState {
+  case initial
+  case requested
+  case downloading
+  case failed(error: Error)
+}
+struct PaginatedOperationState<T, U> {
+  let list: [T]
+  let reload: EmptyRequestState
+  let loadMore: EmptyRequestState
+  let pageInfo: PageInfo
+  
+  enum PageInfo {
+    case initial, lastPage, next(U)
+  }
+}
+
+struct PaginationState: StateType {
+  typealias CategoryState = PaginatedOperationState<Movie.ID, Int>
+  
+  let paginated: [MovieCategory.ID: CategoryState]
+}
+
+// MARK: - PaginationState Reducer -
+extension PaginationState {
+  static func reduce(action: Action, state: PaginationState) -> PaginationState {
+    switch action {
+    case let movieCategoriesAction as MovieCategoriesAction:
+      return reduceIfMovieCategoriesAction(action: movieCategoriesAction, state: state)
+    case let requestAction as MoviesRequestAction:
+      return reduceIfMoviesRequestAction(action: requestAction, state: state)
+    case let responseAction as MoviesResponseAction:
+      return reduceIfMoviesResponseAction(action: responseAction, state: state)
+    default: return state
+    }
+  }
+  // MARK: - MovieCategoriesAction
+  private static func reduceIfMovieCategoriesAction(action: MovieCategoriesAction, state: PaginationState)
+  -> PaginationState {
+    switch action {
+    case let .completed(_,_, relational):
+      return PaginationState(paginated: relational.mapValues { CategoryState.init(list: $0,
+                                                                                  reload: .initial,
+                                                                                  loadMore: .initial,
+                                                                                  pageInfo: .next(2)) })
+    default: return state
+    }
+  }
+  // MARK: - MoviesRequestAction
+  private static func reduceIfMoviesRequestAction(action: MoviesRequestAction, state: PaginationState)
+  -> PaginationState {
+    let categoryId = action.categoryId
+    var newPaginated = state.paginated
+    switch action.requestType {
+    case .reload:
+      newPaginated.updateValue(.init(list: state.paginated[categoryId]?.list ?? [],
```

Comment:
To avoid that, we can implement `Defaultable` protocol for all states.

Defaultable should have only one thing:
`static let defaultValue = ...`

So you will be able to get rid of optional writing it as:
let paginated = state.paginated[categoryId] ?? .defaultValue.

All things that have an initial state can follow that protocol.